### PR TITLE
Fix #4711:Incorrect behavior when empty client id is sent in authorize request

### PR DESCRIPTION
**Pull request by IsurangaPerera**
### Proposed changes in this pull request

- Resolves wso2/product-is#4711

In the current implementation, app-validity and app-state are checked at the time of building the OAuthMessage object and return a JSON string if clientId is an invalid one. Since the error message is seen by the end-user instead of returning a plain JSON string we will redirect the users to an error page.
This is not against the OAuth 2.0 spec as stated in **section 4.2.2.1**

> If the request fails due to a missing, invalid, or mismatching
>    redirection URI, or if the client identifier is missing or invalid,
>    the authorization server SHOULD inform the resource owner of the
>    error and MUST NOT automatically redirect the user-agent to the
>    invalid redirection URI.
> 
>    If the resource owner denies the access request or if the request
>    fails for reasons other than a missing or invalid redirection URI,
>    the authorization server informs the client by adding the following
>    parameters to the fragment component of the redirection URI using the
>    "application/x-www-form-urlencoded" format, per Appendix B: 
> 
**Test Case changes**
- Add test case to handle empty client_id
- Change invalid client Id test case so that client_id and app is validated at the time of app validation
- In case of invalid/empty client_id, user will be redirected to an error page instead of sending a **401** response

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java [hunk 1]
```diff
@@ -130,9 +130,31 @@ public OAuth2ClientValidationResponseDTO validateClientInfo(String clientId, Str
         }
 
         try {
+            if (StringUtils.isBlank(clientId)) {
+                throw new InvalidOAuthClientException("Invalid client_id. No OAuth application has been registered " +
+                        "with the given client_id");
+            }
             OAuthAppDAO oAuthAppDAO = new OAuthAppDAO();
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: We should avoid calling the DAO directly since it's not a cached back DAO. For the moment we have to rely on using the OAuth2Util.getAppInformationByClientId(clientId) method which makes use of the cache.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java [hunk 2]
```diff
@@ -148,7 +170,7 @@ public OAuth2ClientValidationResponseDTO validateClientInfo(String clientId, Str
             OAuth2Util.setClientTenatId(IdentityTenantUtil.getTenantId(appDO.getUser().getTenantDomain()));
 
             // Valid Client, No callback has provided. Use the callback provided during the registration.
-            if (callbackURI == null) {
+            if (StringUtils.isBlank(callbackURI)) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: 1. Not sending the 'redirect_uri' query param is allowed in the spec. 
2. If the redirect_uri query param is sent with an empty value we should consider it as an invalid request.

With this change, both cases will be considered valid. This is not the expected behaviour isn't it?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/state/OAuthRequestStateValidator.java [hunk 3]
```diff
@@ -114,15 +114,6 @@ private void validateRequest(OAuthMessage oAuthMessage)
                 oAuthMessage.setSessionDataKeyFromConsent(null);
             }
         }
-
-        validateOauthApplication(oAuthMessage);
-    }
-
-    private void validateOauthApplication(OAuthMessage oAuthMessage) throws InvalidRequestParentException {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/state/OAuthRequestStateValidator.java]: Any reason why we removed this method?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java (hunk 1 above)
Comment (mefarazath) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: The same class has getOauthApplicationState(String consumerKey) method which does this. Let's reuse that here.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java (hunk 2 above)
Comment (IsurangaPerera) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: Fixed

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java [hunk 4]
```diff
@@ -130,8 +130,29 @@ public OAuth2ClientValidationResponseDTO validateClientInfo(String clientId, Str
...
+            }
+            OAuthAppDO appDO = OAuth2Util.getAppInformationByClientId(clientId);
+
+            String appState = appDO.getState();
+
+            if (StringUtils.isEmpty(appState)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("A valid OAuth client could not be found for client_id: " + clientId);
+                }
+
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: Remove unnecessary new lines.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java [hunk 5]
```diff
@@ -130,8 +130,29 @@ public OAuth2ClientValidationResponseDTO validateClientInfo(String clientId, Str
...
 
         try {
-            OAuthAppDAO oAuthAppDAO = new OAuthAppDAO();
-            OAuthAppDO appDO = oAuthAppDAO.getAppInformation(clientId);
+            if (StringUtils.isBlank(clientId)) {
+                throw new InvalidOAuthClientException("Invalid client_id. No OAuth application has been registered " +
+                        "with the given client_id");
+            }
+            OAuthAppDO appDO = OAuth2Util.getAppInformationByClientId(clientId);
+
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: Remove unnecessary new lines.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java [hunk 6]
```diff
@@ -130,8 +130,29 @@ public OAuth2ClientValidationResponseDTO validateClientInfo(String clientId, Str
         }
 
         try {
-            OAuthAppDAO oAuthAppDAO = new OAuthAppDAO();
-            OAuthAppDO appDO = oAuthAppDAO.getAppInformation(clientId);
+            if (StringUtils.isBlank(clientId)) {
+                throw new InvalidOAuthClientException("Invalid client_id. No OAuth application has been registered " +
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: Is the error msg is correct?

Comment (IsurangaPerera) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: Since the end-user is redirected to the error page, having a more specific error message such as client_id is null or empty is not necessary.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java [hunk 7]
```diff
@@ -130,8 +130,27 @@ public OAuth2ClientValidationResponseDTO validateClientInfo(String clientId, Str
...
+                if (log.isDebugEnabled()) {
+                    log.debug("A valid OAuth client could not be found for client_id: " + clientId);
+                }
+                throw new InvalidOAuthClientException("A valid OAuth client could not be found for client_id: " +
+                        Encode.forHtml(clientId));
+            }
+
+            if (!appState.equalsIgnoreCase(APP_STATE_ACTIVE)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("App is not in active state in client ID: " + clientId + ". App state is:" + appState);
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: Add a space after `state is:`

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/state/OAuthRequestStateValidator.java (hunk 3 above)
Comment (IsurangaPerera) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/state/OAuthRequestStateValidator.java]: With the modifications(as stated in the description) we no longer need this validation method as the clientId validation will be done at the time of app validation.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java (hunk 7 above)
Comment (IsurangaPerera) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: Fixed
